File: [구현] 큰 수 만들기.cpp
#include <string>
#include <vector>
#include <algorithm>
using namespace std;

string solution(string number, int k) {

	string bucket;
	vector<int> removeAt;

	// MARK: - 주어진 숫자로부터 하나씩 꺼내어 모으면서, 이때 모아둔 것 중 지금 등장한 것보다 작은 것들을 빼낸다.
	for (const auto letter : number) {
		// MARK: - 현재 조합하고자 하는 배열이 비어있는 경우
		if (bucket.empty()) { bucket.push_back(letter); continue; }

		// MARK: - 현재의 숫자들이 추가 될 숫자보다 작은 경우에는 제거한다.
		while (!bucket.empty() && removeAt.size() != k && bucket.back() < letter) {
			removeAt.push_back(bucket.back()); bucket.pop_back();
		}
		
		bucket.push_back(letter);
	}

	// MARK: - 내림차순으로 정렬이 되어있는 경우에는 끝의 K만큼 제거한다.
	string answer = removeAt.empty() ? number.substr(0, number.size() - k) : bucket;
	return answer;
}

File: [구현] 큰 수 만들기.swift
import Foundation

/**
    - NOTE:
        - number는 1자리 이상, 1,000,000자리 이하인 숫자입니다.
        - k는 1 이상 number의 자릿수 미만인 자연수입니다.
*/
func solution(_ number: String, _ k: Int) -> String {
    
    var removeRemainder = k
    var result: String = String.init()
    
    // 해당 문제를 구현함에 있어 주의할 점은 해당 숫자의 위치를 움직여서 큰 수를 만드는 방식이 아니다.
    for letter in number {
        
        if result.isEmpty {
            result.append(letter)
            continue
        }
        
        while !result.isEmpty && result.last! < letter && removeRemainder > Int.zero {
            removeRemainder -= 1
            result.removeLast()
        }
        
        result.append(letter)
    }
    
    // MARK: - number에서 k 개의 수를 제거했을 때 만들 수 있는 수 중 가장 큰 숫자를 문자열 형태로 반환하세요.
    if removeRemainder > Int.zero {
        let endIndex = result.index(result.startIndex, offsetBy: number.count - removeRemainder)
        return String(result[result.startIndex..<endIndex])
    }
        
    return result
}
